### DISCUSSION TITLE
Coding query ... if __name__ == '__main__':  and pin hold

### DISCUSSION THREAD

**Discussion by davefes** [micropython/micropython#10396] [Using MicroPython] [answered]
Having discovered:

> if __name__ == '__main__':

I don't understand what happens in this code:
```py
from machine import Pin
import utime


pin32 = Pin(32, Pin.OUT, hold=False)
pin32.off()


def rotate_valve():
    pin32.on()
    print (f'pin32 is on')
    utime.sleep(0.1)


def main():
    global pin32


    while True:
        pin32.off() #  make sure it is off every loop
        print (f'pin32 is off')
        utime.sleep(0.1)

        utime.sleep(1)
        rotate_valve()
        utime.sleep(1)

     #  hold the GPIO state during lightsleep()
        pin32 = Pin(32, Pin.OUT, hold=True) #  motor drive (line in question)


if __name__ == '__main__':
    main()
```
If I don't put `global pin32` at the top of def main(): I get the error:
`Traceback (most recent call last):
  File "main.py", line 33, in <module>
  File "main.py", line 20, in main (line in question)
NameError: local variable referenced before assignment
`
It appears that doing pin32.on() is OK but not OK if trying to change any of its parameters.

A keyword should provide the relief I need.

**Comment by karfas** [comment, marked as answer]
> If I don't put `global pin32` at the top of def main(): I get the error: `Traceback (most recent call last): File "main.py", line 33, in <module> File "main.py", line 20, in main (line in question) NameError: local variable referenced before assignment ` It appears that doing pin32.on() is OK but not OK if trying to change any of its parameters.

`pin32 = Pin(32, Pin.OUT, hold=False)` does not merely change a parameter of `pin32`. For python, this creates a new instance of a `Pin` object with the new parameters, even when this is only a reference to the already-existing Pin object (due to the most-likely singleton characteristic of Pin objects).
After that, the old (maybe global) pin32 instance is gone and away (as seen from inside the function).

You misunderstand how function-local variables get created in python. **Any assignment** to a name (e.g. `pin32 = Pin(...)` in your example) **creates** a variable (local in absence of the `global` declaration).
When **accessing** a variable (`pin32.on()`), python searches for variables with 1) local and 2) global scope, therefore you can **use** any variable created on global level from inside your functions.

**Reply by davefes** [reply]
pin32 = Pin(32, Pin.OUT, hold=False) OK, creating a new instance.  I still don't understand why it needed the global pin32, in order to create this instance without error.

I understand that the above creates a local variable.

> When accessing a variable (pin32.on())

Isn't pin32.on() a function?

Is there something about how functions and local variables behave that I have missed?  Or global and local scope?

I am going to delete "Another fix:" from the original post as that looks incorrect or even more confusing.

Thanks for your reply.

**Reply by sosi-deadeye** [reply]
> I still don't understand why it needed the global pin32, in order to create this instance without error.

This happens because you have pin32 already on module level and if you assign in a function an object to this name, it must be global.

Instead of creating a new instance and assigning it to the same name, reconfigure the pin. Then it is not an assignment, and you can get rid of `global pin32`.

Try this in your function:
```
p32.init(hold=True)
```

**Reply by karfas** [reply]
> > When accessing a variable (pin32.on())
> 
> Isn't pin32.on() a function?
> 
Yes and no. `pin32` is a variable. `pin32.on()` is a method call and therefore a function call. 
However, calling `pin32.on()` requires first to access the `pin32` **variable** to find the methods (if any) of the object (when it's an object) referenced by `pin32`.

**Reply by davefes** [reply]
karfas,

After a late night and turning the computer off, your comment about creating a new object rather than me just wanting to change its behaviour, was the clue that I needed.

If I give the new object a new name the error disappears.
```py
from machine import Pin
import utime

pin32 = Pin(32, Pin.OUT, hold=False)
pin32.off()

def rotate_valve():
    pin32.on()
    print (f'pin32 is on')
    utime.sleep(0.1)

def main():
    while True:
        pin32.off() #  make sure it is off every loop
        print (f'pin32 is off')
        utime.sleep(0.1)

        utime.sleep(1)
        rotate_valve()
        utime.sleep(1)

     #  hold the GPIO state during lightsleep()
        pin32_hold = Pin(32, Pin.OUT, hold=True) #  motor drive (line in question)

if __name__ == '__main__':
    main()
```
**N.B. the above is a kludge!**

If I am going to change the behaviour of pin32 many times in main(), maybe that is a good case for making pin32 global.

Thank you for your persistence.

sosi-deadeye,
Then I noticed your comment.  I tried that ... it make more senses then creating another object.  Thanks for the hint.
https://docs.micropython.org/en/latest/esp32/quickref.html does mention pin.init() and the hold parameter,

Looks like I can't mark yours as the answer.

**Reply by bixb922** [reply]
I would assign pin32 in main and pass as parameter to rotate_valve. No global nor module level variables required, all variables are local.

**Reply by davefes** [reply]
Good point.  Did that all the time in my embedded C programs but I do struggle a bit with many aspects of Python, ie init, self, etc
